### PR TITLE
EN-4028 + EN-2949 Enforcer related fixes for automated testing

### PR DESCRIPTION
**Pull request by pxjohnny**
1) Added support for additional activity handler
2) Added px_cookie_hmac sending

### CODE REVIEW

**Comment by nitzanpx** [comment]
please add new configuration to readme

File: lib/utils/pxapi.js [hunk 1]
```diff
@@ -57,6 +57,7 @@ function callServer(pxCtx, callback) {
     if (uuid) {
         data.uuid = uuid;
     }
+    pxCtx.hasMadeServerCall = true;
```
Comment (nitzanpx) [inline lib/utils/pxapi.js]: add default value on pxCtx, else it will be send undefined

File: lib/perimeterx.js [hunk 2]
```diff
@@ -68,32 +67,37 @@ function pxMiddleware(req, res, next) {
             });
         } else {
             pxutil.verifyUserScore(pxCtx, (action) => {
-                pxLogger.debug('score action ' + action);
-                if (action === pxconfig.SCORE_EVALUATE_ACTION.COOKIE_PASS_TRAFFIC) {
-                    pxLogger.debug('sending page requested activity from cookie');
-                    pxCtx.passReason = pxconfig.PASS_REASON.COOKIE
-                    return pxPass(pxCtx, next);
+                let isTerminated = false;
```
Comment (nitzanpx) [inline lib/perimeterx.js]: for my understanding, `ADDITIONAL_ACTIVITY_HANDLER`  function is just to send activities using an additional handler, why would it return boolean value of terminated?

File: lib/perimeterx.js [hunk 3]
```diff
@@ -50,72 +49,72 @@ function pxMiddleware(req, res, next) {
...
+                if (pxconfig.CUSTOM_REQUEST_HANDLER){
+                  pxconfig.CUSTOM_REQUEST_HANDLER(pxCtx, req, res);
+                  return next();
                 }
 
-                if (pxconfig.MODULE_MODE === pxconfig.MONITOR_MODE.MONITOR) {
-                    pxLogger.debug('monitoring mode is on, passing request');
-                    pxCtx.passReason = pxconfig.PASS_REASON.MONITOR_MODE;
-                    return pxPass(pxCtx, next);
+                if (pxCtx.score < pxconfig.BLOCKING_SCORE || pxconfig.MODULE_MODE === pxconfig ){
```
Comment (yaronpx) [inline lib/perimeterx.js]: `pxconfig.MODULE_MODE === pxconfig`?
I don't think this condition is ever true

File: lib/perimeterx.js [hunk 4]
```diff
@@ -50,72 +49,72 @@ function pxMiddleware(req, res, next) {
...
                 pxCtx.passReason = pxconfig.PASS_REASON.CAPTCHA;
-                return pxPass(pxCtx, next);
+                pxPass(pxCtx);
+                return next();
+
             });
         } else {
             pxutil.verifyUserScore(pxCtx, (action) => {
                 pxLogger.debug('score action ' + action);
-                if (action === pxconfig.SCORE_EVALUATE_ACTION.COOKIE_PASS_TRAFFIC) {
```
Comment (yaronpx) [inline lib/perimeterx.js]: why are you deleting these blocks?

Comment (pxjohnny) [inline lib/perimeterx.js]: they are handled much better now... its all under pxCtx.passReason

File: lib/utils/pxapi.js (hunk 1 above)
Comment (pxjohnny) [inline lib/utils/pxapi.js]: handled in site level

**Comment by nitzanpx** [comment]
please update readme regarding new changes and addons, except that all good
very nice work

**Pull request merged**